airbyte-lib/tests/integration_tests/test_source_faker_integration.py

```diff
@@ -0,0 +1,216 @@
+            "seed": SEED_A,
+            "parallelism": 16,  # Otherwise defaults to 4.
+        },
+        install_if_missing=False,  # Should already be on PATH
+    )
+    source.check()
+    # TODO: We can optionally add back 'users' once Postgres can handle complex object types.
+    source.set_streams([
+        "products",
+        "purchases",
+    ])
+    return source
+
+
+@pytest.fixture(scope="function")  # Each test gets a fresh source-faker instance.
+def source_faker_seed_b() -> ab.Source:
+    """Fixture to return a source-faker connector instance."""
+    source = ab.get_connector(
+        "source-faker",
+        local_executable="source-faker",
+        config={
+            "count": FAKER_SCALE_B,
+            "seed": SEED_B,
+            "parallelism": 16,  # Otherwise defaults to 4.
+        },
+        install_if_missing=False,  # Should already be on PATH
+    )
+    source.check()
+    # TODO: We can optionally add back 'users' once Postgres can handle complex object types.
+    source.set_streams([
+        "products",
+        "purchases",
+    ])
+    return source
+
+
+@pytest.fixture(scope="function")
+def duckdb_cache() -> Generator[caches.DuckDBCache, None, None]:
+    """Fixture to return a fresh cache."""
+    cache: caches.DuckDBCache = ab.new_local_cache()
+    yield cache
+    # TODO: Delete cache DB file after test is complete.
+    return
+
+
+@pytest.fixture(scope="function")
+def snowflake_cache(snowflake_config) -> Generator[caches.SnowflakeCache, None, None]:
+    """Fixture to return a fresh cache."""
+    cache: caches.SnowflakeCache = caches.SnowflakeSQLCache(snowflake_config)
+    yield cache
+    # TODO: Delete cache DB file after test is complete.
+    return
+
+
+@pytest.fixture(scope="function")
+def postgres_cache(new_pg_cache_config) -> Generator[caches.PostgresCache, None, None]:
+    """Fixture to return a fresh cache."""
+    cache: caches.PostgresCache = caches.PostgresCache(config=new_pg_cache_config)
+    yield cache
+    # TODO: Delete cache DB file after test is complete.
+    return
+
+
+@pytest.fixture
+def all_cache_types(
+    duckdb_cache: ab.DuckDBCache,
+    snowflake_cache: ab.SnowflakeCache,
+    postgres_cache: ab.PostgresCache,
+):
+    _ = postgres_cache
+    return [
+        duckdb_cache,
+        postgres_cache,
+        # snowflake_cache,  # Snowflake works, but is slow and expensive to test. # TODO: Re-enable.
+    ]
+
+
+def test_faker_pks(
+    source_faker_seed_a: ab.Source,
+    duckdb_cache: ab.DuckDBCache,
+) -> None:
+    """Test that the append strategy works as expected."""
+
+    catalog: ConfiguredAirbyteCatalog = source_faker_seed_a.configured_catalog
+
+    assert len(catalog.streams) == 2
+    assert catalog.streams[0].primary_key
+    assert catalog.streams[1].primary_key
+
+    read_result = source_faker_seed_a.read(duckdb_cache, write_strategy="append")
+    assert read_result.cache._get_primary_keys("products") == ["id"]
```

Comment:
would be nice to have. a test for the composite primary keys as that logic is never called at the moment